# Documentation for script: _YutanpaNet.py_


## Summary
This program applies a network-based approach to infer complete multicomponent transport
systems in a (meta)genome. It generates 4 tab-delimited files with the inferences:

1) _"greens.tsv"_, contains the assignments with the highest confidence.  

2) _"yello_greens.tsv"_, contains less confidence inferences.  

3) _"refined_reds.tsv"_, from the least confident inferences, this file contains the 
sytems with higher chances of being complete.  

4) _"reds.tsv"_, all systems for which the program could not determine whether the 
systems are present in the genome or not.  

In addition, this program generates the following four network files that are used by script 
[_showsubnet.py_](showsubnet.md) to generate a graphical layout of any subnetwork of 
interest: _raw_network_nodes.txt_, _raw_network_edges.txt_, _recom_network_nodes.txt_, and
_recom_network_edges.txt_.


## Contributors  
Yichi Zhang and Arturo Medrano-Soto  


## Dependencies
The following Python module needs to be available to the script: 

1. **Python 3.X**  
You can download the latest version of Python 3 from the [official website](https://www.python.org/).
YutanpaNet was tested with Python 3.7.4 in MacOS Catalina 10.5.6. The following modules need to be available:  
  a) Biopython 1.75  
  b) json 0.8.5  
  c) networkx 2.3  
  d) more-itertools 7.2.0  
  e) numpy 1.17.2  
  f) pandas 0.25.1  
  g) hmmscanParser (part of this distribution)
  
2. **Output file of program _getMultCompSystems.pl_**  
This program requires the tab-delimited output file of script [_getMultCompSystems.pl_](https://github.com/SaierLaboratory/TCDBtools/blob/master/manuals/getMultCompSystems.md) to generate the network of (meta)genome-TCDB associations.  

3. **_blast+ 2.6.0 to 2.10.0_**  
Other versions of blast may require minor adaptations. Visit the
[download site](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&DOC_TYPE=Download). 


## Command line options
The following options are available:


    usage: YutanpaNet.py [-h] -d <raw data table> -tcdb <sequence file> -ra
                         <directory path> -ft <genome feature table> [-lt] -clan
                         <pfam clans file> [-c] [-dot DOT] [-rcov POG]

    optional arguments: 
     -h, --help         show this help message and exit
     
     -d <raw data table>, --data <raw data table>
                        MANDATORY. Path to the output table in tsv format
                        generated by program getMultCompSystems.pl. This table
                        contains all the multicomponent systems matched by the
                        query genome.
                        
      -tcdb <sequence file>, --tcdb-proteins <sequence file>
                        MANDATORY. Path to the file in fasta format with All
                        the protein content in TCDB as generated by the
                        program extractFamily.pl. This file should reflect the
                        TCDB version that was used to analyze both single and
                        multicomponet systems in the query genome.
                        
      -ra <directory path>, --rootaddress <directory path>
                        MANDATORY. Path to the main output directory generated
                        by program getMultCompSystems.pl. This directory
                        contains the output reports as well as all hydropathy
                        plots for every match between the query genome and
                        multicomponent systems in TCDB.
                        
      -ft <genome feature table>, --featuretable <genome feature table>
                        MANDATORY. Path to the feature table file as
                        downloaded from NCBI genome assemblies. This file is
                        used to extract the genomic context of genes and it
                        must be compressed in gzip format.
                        
      -lt, --locus_tag  Flag. If set, the program will use locus_tag
                        accessions to identify proteins in the query genome.
                        Users are responsible for deciding whether RefSeq or
                        locus_tag accessions will be used in the analysis.
                        Default value is RefSeq.
                        
      -clan <pfam clans file>, --clan_info_file <pfam clans file>
                        MANDATORY. Path to the file with Pfam clans as
                        distributed by Pfam (i.e., Pfam-A.clans.tsv.gz). The
                        information in this file is used to select the best
                        matches reported by program getMultCompSystems.pl.
                        This file should be compressed in gzip format
                        
      -c, --circular    Flag. If set, the replicon structure of the query
                        genome is regarded as circular. This information is
                        used to calculate the distance between pairs of genes.
                        The default setting is: linear.
                        
      -dot DOT, --degree_of_tolerance DOT
                        Integer greater than zero. Internally, the network of
                        relationships among protein candidates, TC components
                        and TC systems is represented as a weighted graph. The
                        program finds the shortest path with the highest
                        possible weight between a candidate transporter in the
                        query genome and a multicomponent system in TCDB. The
                        parameter gives the program a flexiblility to find a
                        range of less possible pairs of candidate and system.
                        The higher the parameter, the greater the flexibility.
                        Default value is 2.
                        
      -rcov POG, --reciprocal_coverage POG
                        The minimum alignment coverage for both the query and
                        subject proteins to be considered candidate
                        homologous. Default value is 0.76.

